File: frontend/src/api/attendance.ts
import { get, post, put, del } from "./index"
import type { PaginatedData, SuccessResponse } from "@/types/api"

export interface AttendanceRuleItem {
  id: string; name: string; department?: string | null; check_in_time: string; check_out_time: string
  work_days?: string[]; late_threshold: number; early_leave_threshold: number; overtime_rate?: number | null
  is_active: boolean; created_at?: string | null; updated_at?: string | null
}
export interface AttendanceRecordItem {
  id: string; employee_id: string; date: string; check_in_time?: string | null; check_out_time?: string | null
  check_in_status: string; check_out_status: string; overtime_hours?: number | null; source: string; remark?: string | null; created_at?: string | null; updated_at?: string | null
}
export interface EmployeeOption { id: string; employee_no: string; name: string; department?: string | null }

export function getAttendanceRules() { return get<AttendanceRuleItem[]>("/attendance/rules") }
export function createAttendanceRule(data: Partial<AttendanceRuleItem>) { return post<AttendanceRuleItem>("/attendance/rules", data) }
export function updateAttendanceRule(id: string, data: Partial<AttendanceRuleItem>) { return put<AttendanceRuleItem>("/attendance/rules/" + id, data) }
export function deleteAttendanceRule(id: string) { return del<SuccessResponse>("/attendance/rules/" + id) }
export function getAttendanceRecords(params: { page?: number; page_size?: number; employee_id?: string; date_from?: string; date_to?: string }) { return get<PaginatedData<AttendanceRecordItem>>("/attendance/records", { params }) }
export function createAttendanceRecord(data: { employee_id: string; date: string; check_in_time?: string | null; check_out_time?: string | null; check_in_status?: string; check_out_status?: string; remark?: string | null }) { return post<AttendanceRecordItem>("/attendance/records", data) }
export function updateAttendanceRecord(id: string, data: Record<string, unknown>) { return put<AttendanceRecordItem>("/attendance/records/" + id, data) }
export function deleteAttendanceRecord(id: string) { return del<SuccessResponse>("/attendance/records/" + id) }
export function getAttendanceEmployees() { return get<EmployeeOption[]>("/attendance/employees") }

File: backend/app/schemas/attendance.py
from pydantic import BaseModel
from datetime import date, datetime, time
from typing import Optional

class AttendanceRuleCreate(BaseModel):
    name: str; department: Optional[str] = None; check_in_time: time; check_out_time: time
    work_days: Optional[list[str]] = None; late_threshold: int = 0; early_leave_threshold: int = 0
    overtime_rate: float = 1.5; is_active: bool = True

class AttendanceRuleUpdate(BaseModel):
    name: Optional[str] = None; department: Optional[str] = None; check_in_time: Optional[time] = None
    check_out_time: Optional[time] = None; work_days: Optional[list[str]] = None
    late_threshold: Optional[int] = None; early_leave_threshold: Optional[int] = None
    overtime_rate: Optional[float] = None; is_active: Optional[bool] = None

class AttendanceRuleResponse(BaseModel):
    id: str; name: str; department: Optional[str] = None; check_in_time: str; check_out_time: str
    work_days: Optional[list] = None; late_threshold: int = 0; early_leave_threshold: int = 0
    overtime_rate: Optional[float] = None; is_active: bool; created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None; model_config = {"from_attributes": True}

class AttendanceRecordCreate(BaseModel):
    employee_id: str; date: date; check_in_time: Optional[datetime] = None; check_out_time: Optional[datetime] = None
    check_in_status: str = "normal"; check_out_status: str = "normal"; source: str = "manual_input"
    remark: Optional[str] = None; overtime_hours: Optional[float] = 0

class AttendanceRecordUpdate(BaseModel):
    check_in_time: Optional[datetime] = None; check_out_time: Optional[datetime] = None
    check_in_status: Optional[str] = None; check_out_status: Optional[str] = None
    remark: Optional[str] = None; overtime_hours: Optional[float] = None

class AttendanceRecordResponse(BaseModel):
    id: str; employee_id: str; date: date; check_in_time: Optional[datetime] = None
    check_out_time: Optional[datetime] = None; check_in_status: str; check_out_status: str
    source: str; remark: Optional[str] = None; overtime_hours: Optional[float] = None; created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None; model_config = {"from_attributes": True}